SignalServiceKit/src/Util/OWSTypingIndicators.swift

```diff
@@ -0,0 +1,333 @@
+
+    @objc
+    func didReceiveTypingStartedMessage(inThread thread: TSThread, recipientId: String, deviceId: UInt)
+
+    @objc
+    func didReceiveTypingStoppedMessage(inThread thread: TSThread, recipientId: String, deviceId: UInt)
+
+    @objc
+    func didReceiveIncomingMessage(inThread thread: TSThread, recipientId: String, deviceId: UInt)
+
+    // TODO: Use this method.
+    @objc
+    func areTypingIndicatorsVisible(inThread thread: TSThread, recipientId: String) -> Bool
+}
+
+// MARK: -
+
+@objc
+public class OWSTypingIndicatorsImpl: NSObject, OWSTypingIndicators {
+    @objc public static let typingIndicatorStateDidChange = Notification.Name("typingIndicatorStateDidChange")
+
+    @objc
+    public func inputWasTyped(inThread thread: TSThread) {
+        AssertIsOnMainThread()
+        guard let outgoingIndicators = ensureOutgoingIndicators(forThread: thread) else {
+            owsFailDebug("Could not locate outgoing indicators state")
+            return
+        }
+        outgoingIndicators.inputWasTyped()
+    }
+
+    @objc
+    public func messageWasSent(inThread thread: TSThread) {
+        AssertIsOnMainThread()
+        guard let outgoingIndicators = ensureOutgoingIndicators(forThread: thread) else {
+            owsFailDebug("Could not locate outgoing indicators state")
+            return
+        }
+        outgoingIndicators.messageWasSent()
+    }
+
+    @objc
+    public func didReceiveTypingStartedMessage(inThread thread: TSThread, recipientId: String, deviceId: UInt) {
+        AssertIsOnMainThread()
+        let incomingIndicators = ensureIncomingIndicators(forThread: thread, recipientId: recipientId, deviceId: deviceId)
+        incomingIndicators.didReceiveTypingStartedMessage()
+    }
+
+    @objc
+    public func didReceiveTypingStoppedMessage(inThread thread: TSThread, recipientId: String, deviceId: UInt) {
+        AssertIsOnMainThread()
+        let incomingIndicators = ensureIncomingIndicators(forThread: thread, recipientId: recipientId, deviceId: deviceId)
+        incomingIndicators.didReceiveTypingStoppedMessage()
+    }
+
+    @objc
+    public func didReceiveIncomingMessage(inThread thread: TSThread, recipientId: String, deviceId: UInt) {
+        AssertIsOnMainThread()
+        let incomingIndicators = ensureIncomingIndicators(forThread: thread, recipientId: recipientId, deviceId: deviceId)
+        incomingIndicators.didReceiveIncomingMessage()
+    }
+
+    @objc
+    public func areTypingIndicatorsVisible(inThread thread: TSThread, recipientId: String) -> Bool {
+        AssertIsOnMainThread()
+
+        let key = incomingIndicatorsKey(forThread: thread, recipientId: recipientId)
+        guard let deviceMap = incomingIndicatorsMap[key] else {
+            return false
+        }
+        for incomingIndicators in deviceMap.values {
+            if incomingIndicators.isTyping {
+                return true
+            }
+        }
+        return false
```

Comment:
This method will likely be replaced by a method which can be used to determine "which recipient (if any) should we show the typing indicator for in thread X?".